[Cocoa] Expose Objective-C wrappers for LinearMediaKit types

aestes opened the pull request:
#### 2b5fd9cfa4d56becadf0ed628af6e62ddbf25362
<pre>
[Cocoa] Expose Objective-C wrappers for LinearMediaKit types
<a href="https://bugs.webkit.org/show_bug.cgi?id=268341">https://bugs.webkit.org/show_bug.cgi?id=268341</a>
<a href="https://rdar.apple.com/121886828">rdar://121886828</a>

Reviewed by Richard Robinson and Jer Noble.

LinearMediaKit exposes a Swift SPI that needs to be called by Objective-C++ code in WebKit. To
support this, added Objective-C wrapper types to WebKitSwift that are backed by a Swift
implementation that calls into LinearMediaKit.

Added WKSLinearMediaTypes.h that declares NS_ENUMs, NS_OPTIONs, and @interfaces prefixed with WKS.
For the @interfaces, they are backed by an @_objcImplementation extension in Swift. The enums and
options are also extended with convenience initializers and properties for accessing their
LinearMediaKit equivalents (to be used in follow-on PRs).

* Source/WebKit/Configurations/WebKitSwift.xcconfig:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaTypes.swift: Added.
(WKSLinearMediaTimeRange.lowerBound):
(WKSLinearMediaContentMode.contentMode):
(WKSLinearMediaContentType.contentType):
(WKSLinearMediaPresentationMode.presentationMode):
(WKSLinearMediaViewingMode.viewingMode):
(WKSLinearMediaFullscreenBehaviors.fullscreenBehaviors):
(WKSLinearMediaTimeRange.closedRange):
(WKSLinearMediaTimeRange.range):
* Source/WebKit/WebKitSwift/LinearMediaKit/WKSLinearMediaTypes.h: Added.
* Source/WebKit/WebKitSwift/WebKitSwift.h: Added.
* Source/WebKit/WebKitSwift/module.modulemap: Added.

Canonical link: <a href="https://commits.webkit.org/273728@main">https://commits.webkit.org/273728@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f1c68686eb2ff5a9a2ddbce276f3354ada207cd7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36494 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15433 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38713 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39159 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/32743 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/17899 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12515 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/31374 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37054 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13003 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/32299 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11385 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/11420 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40405 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33070 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32884 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/37344 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11646 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/9505 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/35447 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/13344 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8267 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/12085 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12526 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaTypes.swift:
```diff
@@ -0,0 +1,188 @@
+    var contentMode: ContentMode? {
+        switch self {
+        case .none:
+            return nil
```
can you omit the `return`s?

aestes replied:
I could I guess. I don't know what our style preference is here.

rr-codes replied:
I think at this point we can choose what we want it to be 😂 I personally prefer always omitting the return when possible

rr-codes commented on Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaTypes.swift:
```diff
@@ -0,0 +1,188 @@
+    }
+}
+
+extension WKSLinearMediaTrack: @retroactive Track {
```
what the heck does retroactive do

aestes replied:
It silences this warning:
```
Extension declares a conformance of imported type 'WKSLinearMediaTrack' to imported protocol 'Track'; this will not behave correctly if the owners of 'WebKitSwift' introduce this conformance in the future
```

... which is a silly warning, because we are the actual owners of the WebKitSwift clang module (i.e. the stuff in WebKitSwift.h).

rr-codes replied:
oh wow, cool

rr-codes commented on Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaTypes.swift:
```diff
@@ -0,0 +1,188 @@
+// MARK: Objective-C Implementations
+
+@_objcImplementation extension WKSLinearMediaTimeRange {
+    var lowerBound: TimeInterval
```
why is this var but the other is let?

aestes replied:
whoops, no good reason. will fix.

rr-codes commented on Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaTypes.swift:
```diff
@@ -0,0 +1,188 @@
+            self = .scaleToFill
+        @unknown default:
+            fatalError()
+        }
```
```suggestion
        self = switch contentMode {
        case .scaleAspectFit?:
            .scaleAspectFit
        case .scaleAspectFill?:
           .scaleAspectFill
        case .scaleToFill?:
            .scaleToFill
         case .none:
            .none
        @unknown default:
            fatalError()
        }
```
maybe?

aestes replied:
whoa, cool

aestes replied:
Darn, unfortunately this syntax isn't supported by the swift that ships on Monterey:
```
/Volumes/Data/worker/macOS-Monterey-Release-Build-EWS/build/Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaTypes.swift:128:16: error: expected expression in assignment
        self = switch viewingMode {
```

aestes replied:
Actually, why is this code even building on macOS Monterey? Why would `canImport(LinearMediaKit)` be true there? 🤔

rr-codes commented on Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaTypes.swift:
```diff
@@ -0,0 +1,188 @@
+        case .none:
+            return nil
+        case .mono:
+            return ViewingMode.mono
```
does this need the `ViewingMode` qualifier?

aestes replied:
nope, will fix

aestes committed:
[Cocoa] Expose Objective-C wrappers for LinearMediaKit types
https://bugs.webkit.org/show_bug.cgi?id=268341
rdar://121886828

Reviewed by Richard Robinson and Jer Noble.

LinearMediaKit exposes a Swift SPI that needs to be called by Objective-C++ code in WebKit. To
support this, added Objective-C wrapper types to WebKitSwift that are backed by a Swift
implementation that calls into LinearMediaKit.

Added WKSLinearMediaTypes.h that declares NS_ENUMs, NS_OPTIONs, and @interfaces prefixed with WKS.
For the @interfaces, they are backed by an @_objcImplementation extension in Swift. The enums and
options are also extended with convenience initializers and properties for accessing their
LinearMediaKit equivalents (to be used in follow-on PRs).

* Source/WebKit/Configurations/WebKitSwift.xcconfig:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaTypes.swift: Added.
(WKSLinearMediaTimeRange.lowerBound):
(WKSLinearMediaContentMode.contentMode):
(WKSLinearMediaContentType.contentType):
(WKSLinearMediaPresentationMode.presentationMode):
(WKSLinearMediaViewingMode.viewingMode):
(WKSLinearMediaFullscreenBehaviors.fullscreenBehaviors):
(WKSLinearMediaTimeRange.closedRange):
(WKSLinearMediaTimeRange.range):
* Source/WebKit/WebKitSwift/LinearMediaKit/WKSLinearMediaTypes.h: Added.
* Source/WebKit/WebKitSwift/WebKitSwift.h: Added.
* Source/WebKit/WebKitSwift/module.modulemap: Added.

Canonical link: https://commits.webkit.org/273728@main
